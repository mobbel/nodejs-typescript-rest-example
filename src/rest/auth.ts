import {
  IncomingHttpHeaders,
} from 'http';

const auth = (headers: IncomingHttpHeaders) => {
  if (headers && headers.authorization) {
    // implement here your authorizsation check
    return true;
  }
  return false;
};

export default (auth);
