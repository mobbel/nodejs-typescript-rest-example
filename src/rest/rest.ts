import {
  IncomingHttpHeaders,
} from 'http';
import auth from './auth';

export interface IRest {
  addCallback: (method: string, slug: string, callback, guarded?: boolean) => void;
  render: (method: string, slug: string, requestBody: IRequestBody, headers: IncomingHttpHeaders) => IResponse;
}

interface IResponseHeader {
  headerKey: string;
  headerValue: string;
}

export interface IResponse {
  status: number;
  contentType: string;
  headers?: IResponseHeader[];
  body?: string;
}

export interface IRequestBody {
  [key: string]: string;
}

export interface ISlugObject {
  guarded: boolean;
  callback: (properties: ICallbackProperties) => IResponse;
}

export interface IRegexObject {
  guarded: boolean;
  regex: string;
  regexNames: string[];
  callback: (properties: ICallbackProperties) => IResponse;
}

export interface ICallbackProperties {
  requestBody?: IRequestBody;
  slugVariables?: { [key: string]: string };
}

interface IVariablesFromSlug {
  [key: string]: string;
}

const rest = (): IRest => {
  const callbackHolder = {
    delete: [],
    get: [],
    patch: [],
    post: [],
    put: [],
  };
  const regexCallbackHolder = {
    delete: [],
    get: [],
    patch: [],
    post: [],
    put: [],
  };
  const addCallback = (method: string, slug: string, callback, guarded: boolean = false): void => {
    if (slug.indexOf('/:') < 0) {
      const slugObject: ISlugObject = {
        guarded,
        callback,
      };
      callbackHolder[method][slug] = slugObject;
    } else {
      const slugTempArray = [];
      const regexNames = [];
      const slugArray = slug.split('/');
      slugArray.map((slugToTest) => {
        if (slugToTest[0] !== ':') {
          slugTempArray.push(slugToTest);
        } else {
          slugTempArray.push('([a-z0-9]*)');
          regexNames.push(slugToTest.slice(1));
        }
      });
      const regexObject: IRegexObject = {
        guarded,
        callback,
        regexNames,
        regex: `${slugTempArray.join('\/')}`,
      };
      regexCallbackHolder[method].push(regexObject);
    }
  };

  const render = (
    method: string,
    slug: string,
    requestBody: IRequestBody,
    headers?: IncomingHttpHeaders,
  ) => {
    let returnCode: IResponse;
    if (callbackHolder[method][slug]) {
      if (!callbackHolder[method][slug].guarded) {
        returnCode = callbackHolder[method][slug].callback({ requestBody });
      } else if (callbackHolder[method][slug].guarded && auth(headers)) {
        returnCode = callbackHolder[method][slug].callback({ requestBody });
      } else {
        returnCode = {
          status: 401,
          contentType: 'Unauthorized',
          headers: [
            {
              headerKey: 'www-authorization',
              headerValue: 'missing login status',
            },
          ],
        };
      }
    }

    if (typeof returnCode === 'undefined') {
      regexCallbackHolder[method].map((methodRegex: IRegexObject) => {
        const matches = slug.match(methodRegex.regex);
        console.log(matches);
        if (matches && matches.length > 0 && slug === matches[0]) {
          if (!methodRegex.guarded || (methodRegex.guarded && auth(headers))) {
            const variablesFromSlug: IVariablesFromSlug = {};
            methodRegex.regexNames.map((regexName: string, index: number) => {
              variablesFromSlug[regexName] = matches[index + 1];
            });
            returnCode = (methodRegex.callback)({ requestBody, slugVariables: variablesFromSlug });
          } else {
            returnCode = {
              status: 401,
              contentType: 'Unauthorized',
              headers: [
                {
                  headerKey: 'www-authorization',
                  headerValue: 'missing login status',
                },
              ],
            };
          }
        }
      });
    }

    if (typeof returnCode !== 'undefined') {
      return (returnCode);
    }
    return ({
      status: 404,
      contentType: 'text/plain',
    });
  };

  return {
    addCallback,
    render,
  };
};

export default rest;
