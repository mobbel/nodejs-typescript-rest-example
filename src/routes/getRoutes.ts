import { IRest, ICallbackProperties, IResponse } from '../rest/rest';
import test from '../controller/test';

const getRoutes = (rest: IRest) => {
  rest.addCallback('get', 'test', (): IResponse => {
    return test().testData();
  });
  rest.addCallback('get', 'test/:id', (properties: ICallbackProperties): IResponse => {
    return test().testSlugData(properties);
  });
  rest.addCallback('get', 'test/guarded', (): IResponse => {
    return test().testData();
  },               true);
};

export default getRoutes;
