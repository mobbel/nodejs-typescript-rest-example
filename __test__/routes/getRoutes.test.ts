import rest, { IRest } from '../../src/rest/rest';
import getRoutes from '../../src/routes/getRoutes';

describe('>>> Testing Get-Routes', () => {
  const restApp: IRest = rest();
  getRoutes(restApp);

  test('+++ Snapshot of /test route', () => {
    expect(restApp.render('get', 'test', {}, {})).toMatchSnapshot();
  });

  test('+++ Snapshot of /test/:id route', () => {
    expect(restApp.render('get', 'test/12345', {}, {})).toMatchSnapshot();
  });

  test('+++ Snapshot of /test/guarded fail', () => {
    expect(restApp.render('get', 'test/guarded', {}, {})).toMatchSnapshot();
  });

  test('+++ Snapshot of /test/guarded work', () => {
    expect(restApp.render('get', 'test/guarded', {}, {
      authorization: 'test',
    })).toMatchSnapshot();
  });
});
