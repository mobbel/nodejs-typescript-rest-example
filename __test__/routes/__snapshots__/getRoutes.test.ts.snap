// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> Testing Get-Routes +++ Snapshot of /test route 1`] = `
Object {
  "body": "{\\"test\\":\\"data\\"}",
  "contentType": "application/json",
  "status": 200,
}
`;

exports[`>>> Testing Get-Routes +++ Snapshot of /test/:id route 1`] = `
Object {
  "body": "{\\"slugData\\":{\\"id\\":\\"12345\\"}}",
  "contentType": "application/json",
  "status": 200,
}
`;

exports[`>>> Testing Get-Routes +++ Snapshot of /test/guarded fail 1`] = `
Object {
  "contentType": "Unauthorized",
  "headers": Array [
    Object {
      "headerKey": "www-authorization",
      "headerValue": "missing login status",
    },
  ],
  "status": 401,
}
`;

exports[`>>> Testing Get-Routes +++ Snapshot of /test/guarded work 1`] = `
Object {
  "body": "{\\"test\\":\\"data\\"}",
  "contentType": "application/json",
  "status": 200,
}
`;
